src/borg/_hashindex.c

```diff
@@ -745,16 +748,16 @@ hashindex_set(HashIndex *index, const unsigned char *key, const void *value)
     uint8_t *ptr;
     if(idx < 0)
     {
-        if(index->num_entries > index->upper_limit) {
-            /* hashtable too full, grow it! */
+        if(index->num_entries > index->upper_limit || idx == -2) {
```

Comment:
Since we'll be adding 1 to num_entries below, the condition should be `>=`:
```suggestion
        if(index->num_entries >= index->upper_limit || idx == -2) {
```